File: hash_advanced/08_certification/python3/certification.py
# これはpaizaラーニングの「ハッシュメニュー応用編」-「ログイン認証をしてみよう」に
# Python3でチャレンジしたコードです。
# https://paiza.jp/works/mondai/hash_advanced/hash_advanced__certification
# 
# 作成環境
# Ubuntu 22.04.5 LTS
# Python 3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0]

# ハッシュ関数
# H(p) = (p の 1 文字目の文字コード * B^1 + p の 2 文字目の文字コード * B^2 + ...
#  + p の m 文字目の文字コード * B^m) % mod
# B = 10^8+7, mod = 10^9+7, 文字コードは ASCII
# parameters
#   text 文字列
# return ハッシュ値(数値)
def hashFunction(text):
    B = 10 ** 8 + 7
    mod = 10 ** 9 + 7
    total = 0
    for i in range(len(text)):
        total += ord(text[i]) * B ** (i + 1) % mod
    return total % mod

# ログイン認証をしてみる

# 1 行目に組の数を表す整数 N と M が与えられます。
N, M = map(int, input().split(' '))
# i + 1 行目にデータベースで保管しているアカウント名を表す文字列 A_i と
# それに対応するパスワードのハッシュ値を表す整数 h_i が与えられます。
# (1 ≦ i ≦ N)
data = {} # アカウント名をキーとしてハッシュ値を保存する辞書
for i in range(N):
    A, h = input().split(' ')
    data[A] = int(h) # 数値文字列を数値に変換
# print(data)
# j + N + 1 行目にログイン認証を行うアカウント名を表す文字列 B_j と
# パスワードを表す文字列 p_j が与えられます。(1 ≦ j ≦ M)
for j in range(M):
    B, p = input().split(' ')
    hp = hashFunction(p)
    # print(hp)
    # アカウントが辞書にあるかチェック
    if B in data and data[B] == hp:
        print('Yes')
    else:
        print('No')

File: hash_advanced/08_certification/swift/certification.swift
/*
これはpaizaラーニングの「ハッシュメニュー応用編」-「ログイン認証をしてみよう」に
swiftでチャレンジしたコードです。
https://paiza.jp/works/mondai/hash_advanced/hash_advanced__certification

作成環境
Ubuntu 22.04.5 LTS
Swift version 5.4 (swift-5.4-RELEASE) Target: x86_64-unknown-linux-gnu
*/
import Foundation

// ハッシュ関数
// H(p) = (p の 1 文字目の文字コード * B^1 + p の 2 文字目の文字コード * B^2 + ... +
//  p の m 文字目の文字コード * B^m) % mod
// B = 10^8+7, mod = 10^9+7, 文字コード ASCII
func hashFunction(_ text: String) -> Int {
    // ハッシュ関数で使用する定数
    let B: Int = Int(pow(10.0, 8.0)) + 7
    let mod: Int = Int(pow(10.0, 9.0)) + 7

    var total: Int = 0
    var power: Int = 1
    for char in text {
        power *= B
        power %= mod
        total += Int(char.asciiValue ?? 0) * power % mod
    }
    let h = total % mod
    return h
}

// ログイン認証をしてみる
func main() {
    // 1 行目に組の数を表す整数 N と M が与えられます。
    let items = readLine()!.trimmingCharacters(in: .whitespacesAndNewlines)
                .split(separator: " ")
    let N = Int(items[0]) ?? 0
    let M = Int(items[1]) ?? 0
    // アカウント名をキーとしてハッシュ値を保存する辞書
    var data: [String: Int] = [:]
    // i + 1 行目にデータベースで保管しているアカウント名を表す文字列 A_i と
    // それに対応するパスワードのハッシュ値を表す整数 h_i が与えられます。
    // (1 ≦ i ≦ N)
    for _ in 0 ..< N {
        let items2 = readLine()!.trimmingCharacters(in: .whitespacesAndNewlines)
                    .split(separator: " ")
        data[String(items2[0])] = Int(items2[1]) ?? 0
    }
    // j + N + 1 行目にログイン認証を行うアカウント名を表す文字列 B_j と
    // パスワードを表す文字列 p_j が与えられます。(1 ≦ j ≦ M)
    for _ in 0 ..< M {
        let items3 = readLine()!.trimmingCharacters(in: .whitespacesAndNewlines)
                    .split(separator: " ")
        let h = hashFunction(String(items3[1]))
        if data.keys.contains(String(items3[0])) && data[String(items3[0])] == h {
            print("Yes")
        } else {
            print("No")
        }
    }
}

// エントリーポイント
main()
